### PR TITLE
Add nullable annotations to middleware

### PR DESCRIPTION
**Pull request by pranavkm**
* [x] CORS
* [x] Diagnostics (public types only)
* [x] HeaderPropagation
* [x] HealthChecks
* [x] HostFiltering
* [x] HttpOverrides
* [x] HttpsPolicy
* [x] Localization
* [x] MiddlewareAnalysis
* [x] ResponseCaching

Contributes to https://github.com/dotnet/aspnetcore/issues/5680

### CODE REVIEW

File: src/Http/Http.Features/src/FeatureCollection.cs
Comment (pranavkm) [inline src/Http/Http.Features/src/FeatureCollection.cs:103]: @JamesNK while the compiler used for `dotnet build` supports `TFeature?`, VS 16.8 P2 is required to get the correct design time experience. I'm going to hold off on updating this until that becomes available.

File: src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerMiddleware.cs
Comment (pranavkm) [reply src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerMiddleware.cs:121]: It's initialized as part of the ctor: https://github.com/dotnet/aspnetcore/blob/master/src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerMiddleware.cs#L36-L46

File: src/Middleware/HostFiltering/src/HostFilteringOptions.cs
Comment (pranavkm) [reply src/Middleware/HostFiltering/src/HostFilteringOptions.cs:26]: There's apparently some scenario where it's allowed to be configured as null: https://github.com/dotnet/aspnetcore/blob/4e44025a52e4b73aa17e09a8041b0e166e0c5ce0/src/Middleware/HostFiltering/sample/Startup.cs#L35. It seems unusual, but I didn't want to change the semantics in this PR

File: src/Http/Http.Abstractions/src/Extensions/UseMiddlewareExtensions.cs
Comment (pranavkm) [reply src/Http/Http.Abstractions/src/Extensions/UseMiddlewareExtensions.cs:36]: It's an empty array by default.

File: src/Http/Http.Abstractions/src/Extensions/UseMiddlewareExtensions.cs
Comment (pranavkm) [reply src/Http/Http.Abstractions/src/Extensions/UseMiddlewareExtensions.cs:48]: The ctor args could be null, no? We don't require them to be non-null.

File: src/Middleware/Diagnostics.Abstractions/src/IExceptionHandlerPathFeature.cs
Comment (pranavkm) [reply src/Middleware/Diagnostics.Abstractions/src/IExceptionHandlerPathFeature.cs:15]: Do we need to tweak this: https://github.com/dotnet/aspnetcore/blob/master/src/Http/Http.Abstractions/src/PathString.cs#L42?

File: src/Http/Http.Abstractions/src/PathString.cs
Comment (pranavkm) [inline src/Http/Http.Abstractions/src/PathString.cs:42]: I think this is a better way to go with `PathString`. While you could initialize it with a null `value`, it's ordinarily never null.

File: src/Http/Http.Features/src/FeatureCollection.cs
Comment (pranavkm) [inline src/Http/Http.Features/src/FeatureCollection.cs:97]: @sharwell more of these 👍

File: src/Http/Http.Abstractions/src/Extensions/UsePathBaseExtensions.cs
Comment (sharwell) [inline src/Http/Http.Abstractions/src/Extensions/UsePathBaseExtensions.cs:29]: This one doesn't look right. It's using this operator:

https://github.com/dotnet/aspnetcore/blob/4e01bbe4cfeb8d63cf5ac40fcf1a6a7c77c7e911/src/Http/Http.Abstractions/src/PathString.cs#L460-L461

File: src/Http/Http.Abstractions/src/PathString.cs
Comment (sharwell) [reply src/Http/Http.Abstractions/src/PathString.cs:42]: I'd leave this type with the original definition, but update `HasValue` to have `[MemberNotNullWhen(true, nameof(Value))]`.

File: src/Http/Http.Abstractions/src/PathString.cs
Comment (sharwell) [inline src/Http/Http.Abstractions/src/PathString.cs:23]: ⚠️ This change doesn't handle `default(PathString)` as well as the original code. I'd use the new C# 9 `MemberNotNullWhenAttribute` instead.

File: src/Http/Http.Abstractions/src/QueryString.cs
Comment (sharwell) [inline src/Http/Http.Abstractions/src/QueryString.cs:39]: ⚠️ The original code was better here.

File: src/Http/Http.Features/src/FeatureReferences.cs
Comment (sharwell) [inline src/Http/Http.Features/src/FeatureReferences.cs:67]: 📝 There are some suppressions (`!` operators) in the implementation of this method which can now be removed.

File: src/Http/Http.Features/src/FeatureReferences.cs
Comment (sharwell) [inline src/Http/Http.Features/src/FeatureReferences.cs:83]: ❔ Is it really necessary to allow `factory` to return null here?

File: src/Http/Http/src/Internal/DefaultHttpResponse.cs
Comment (sharwell) [inline src/Http/Http/src/Internal/DefaultHttpResponse.cs:72]: 💡 In Roslyn, we use a `GetRequired<TFeature>` extension method which calls `Get<TFeature>` and throws a meaningful exception if the required feature is not available. I would recommend using that approach instead of just the `!` operator.

File: src/Http/Http/src/Internal/ResponseCookies.cs
Comment (sharwell) [inline src/Http/Http/src/Internal/ResponseCookies.cs:31]: 💡 Another case where `GetRequired` would help

```suggestion
            Headers = _features.GetRequired<IHttpResponseFeature>().Headers;
```

File: src/Middleware/CORS/src/Infrastructure/DefaultCorsPolicyProvider.cs
Comment (sharwell) [inline src/Middleware/CORS/src/Infrastructure/DefaultCorsPolicyProvider.cs:37]: 💡 An [`AsNullable`](https://github.com/dotnet/roslyn/blob/adaa46d777e4be3c70c5f7455e39b58b0db0fb89/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/SpecializedTasks.cs#L27-L28) extension method helps avoid accidental misinterpretation of this suppression.

```suggestion
                return result.policyTask.AsNullable();
```

File: src/Middleware/Diagnostics/src/DeveloperExceptionPage/Views/ErrorPageModel.cs
Comment (sharwell) [inline src/Middleware/Diagnostics/src/DeveloperExceptionPage/Views/ErrorPageModel.cs:4]: 💡 A comment might be helpful

File: src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerFeature.cs
Comment (sharwell) [inline src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerFeature.cs:12]: 💡 Use `null` instead of `default` for reference types

```suggestion
        public Exception Error { get; set; } = null!;
```

File: src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerFeature.cs
Comment (sharwell) [inline src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerFeature.cs:10]: 💡 Consider modifying this to have a constructor instead of just read/write properties.

File: src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerMiddleware.cs
Comment (sharwell) [inline src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerMiddleware.cs:121]: 💡 Consider adding a property (even if private) for the class which returns `_options.ExceptionHandler` as a known non-null value

File: src/Middleware/Diagnostics/src/StatusCodePage/StatusCodeReExecuteFeature.cs
Comment (sharwell) [inline src/Middleware/Diagnostics/src/StatusCodePage/StatusCodeReExecuteFeature.cs:10]: 💡 Consider initialization with a constructor here.

```suggestion
        public string OriginalPath { get; set; } = null!;
```

File: src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs
Comment (sharwell) [inline src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs:84]: 💡 If you create a property `AllowAnyNonEmptyHost`, you can apply the `MemberNotNullWhenAttribute` to it to avoid needing this assertion. Another option is to make `_allowedHosts` never be null (I believe you can assign `Array.Empty<StringSegment>()` to it for the default case without changing the behavior).

File: src/Middleware/Localization/src/RequestCultureProvider.cs
Comment (sharwell) [inline src/Middleware/Localization/src/RequestCultureProvider.cs:19]: 💡 I've seen this a couple times. If there is a shared library where a `SpecializedTasks` class could live, consider something like this:

https://github.com/dotnet/roslyn/blob/4e0ac646bfa6acf43ee165f7c46d4d3a46c0e719/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/SpecializedTasks.cs#L31-L32

File: src/Middleware/Localization/src/RequestLocalizationMiddleware.cs
Comment (sharwell) [inline src/Middleware/Localization/src/RequestLocalizationMiddleware.cs:114]: 💭 This seems to be a slight change in behavior, but I'm not sure if it's a problem or not.

File: src/Middleware/MiddlewareAnalysis/src/AnalysisBuilder.cs
Comment (sharwell) [inline src/Middleware/MiddlewareAnalysis/src/AnalysisBuilder.cs:53]: ⚠️ The new code doesn't appear to adhere to this comment. Possible 🐛

**Comment by shirhatti** [comment]
@pranavkm Do you still plan to address to get to this in 5.0?

File: src/Http/Http.Abstractions/src/PathString.cs
Comment (pranavkm) [reply src/Http/Http.Abstractions/src/PathString.cs:42]: While I agree it's possible for `Value` to be null, in the most common use-cases that users would encounter a `PathString` this would never be null and makes the calling code more cumbersome. In our codebase, we've attempted to model the most common case rather than the exceptional. That said, @Tratcher would know more about how common of an occurence a `null` PathString is.

Comment (Tratcher) [reply src/Http/Http.Abstractions/src/PathString.cs:42]: PathString is just a wrapper type, it's likelyhood of being null is contextual. They type itself should say yes it can be null, and we'd override that in specific contexts like Request.Path which never should be null.

**Comment by pranavkm** [comment]
@shirhatti depends on if we can get this reviewed before Monday.

File: src/Http/Http.Abstractions/src/Extensions/UsePathBaseExtensions.cs
Comment (Tratcher) [reply src/Http/Http.Abstractions/src/Extensions/UsePathBaseExtensions.cs:29]: Oh, that's a bug. Value is already un-escaped, and ConvertFromString would un-escape it again. Double un-escaping is problematic in some contexts. This should be:
```suggestion
            pathBase = new PathString(pathBase.Value?.TrimEnd('/'));
```

File: src/Http/Http.Abstractions/src/QueryString.cs
Comment (Tratcher) [reply src/Http/Http.Abstractions/src/QueryString.cs:39]: Yes, revert this.

File: src/Http/Http/src/Internal/DefaultHttpResponse.cs
Comment (Tratcher) [reply src/Http/Http/src/Internal/DefaultHttpResponse.cs:72]: Agreed. We've been lax in assuming that the request feature, the response feature, and the response body feature would always be set, and we'd break horribly if they aren't.

**Review by Tratcher — CHANGES_REQUESTED** [review]

File: src/Http/Http.Features/src/FeatureReferences.cs
Comment (pranavkm) [reply src/Http/Http.Features/src/FeatureReferences.cs:83]: There's quite a few places that pass a use a factory that produces nulls.

File: src/Http/Http/src/Internal/DefaultHttpResponse.cs
Comment (pranavkm) [reply src/Http/Http/src/Internal/DefaultHttpResponse.cs:72]: I filed https://github.com/dotnet/aspnetcore/issues/25471 so we can take this thru API review.

File: src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerFeature.cs
Comment (pranavkm) [reply src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerFeature.cs:12]: We've been using `default!` all over the place. Is that a preference or is there to more to using `null`?

Comment (sharwell) [reply src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerFeature.cs:12]: `default` suggests that the type is a value type or an unconstrained generic. `null` suggests that the type is a reference type or `Nullable<T>`.

File: src/Middleware/Diagnostics/src/DeveloperExceptionPage/Views/ErrorPageModel.cs
Comment (pranavkm) [reply src/Middleware/Diagnostics/src/DeveloperExceptionPage/Views/ErrorPageModel.cs:4]: We have a number of these internal types that I haven't been annotating in this pass and that's well understood internally. We'll look at updating these piecemeal.

File: src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerFeature.cs
Comment (pranavkm) [reply src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerFeature.cs:10]: It's going to be difficult to consider add a new public API to support nullability.

File: src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerFeature.cs
Comment (pranavkm) [reply src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerFeature.cs:12]: It'd be inconsistent to have it use `default!` in some places but not the other. I'll leave it to someone to bulk update the repo to use `null!` exclusively.

File: src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerMiddleware.cs
Comment (pranavkm) [reply src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerMiddleware.cs:121]: I tried doing this, but it felt like it made the options type more complex for minimal benefit. I appreciate the feedback but I'll leave this one be.

File: src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs
Comment (pranavkm) [reply src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs:84]: Hmm, wouldn't I have to move the assertion in to the `AllowAnyNonEmptyHost` property at that point? Perhaps I'm missing how we would tie to together.

File: src/Middleware/Localization/src/RequestLocalizationMiddleware.cs
Comment (pranavkm) [reply src/Middleware/Localization/src/RequestLocalizationMiddleware.cs:114]: Line 107 asserts that the two cannot be simultaneously null. The null checks inside the if reiterate these.

File: src/Middleware/Localization/src/RequestCultureProvider.cs
Comment (pranavkm) [reply src/Middleware/Localization/src/RequestCultureProvider.cs:19]: Thanks for the suggestion. It's not a lot of code, so I'll keep this until someone complains.

**Review by pranavkm — COMMENTED** [review]
This should be updated and now targeting 6.0

**Review by Tratcher — APPROVED** [review]

**Pull request merged**
